lib/saml/auth_fail_handler.rb

```diff
@@ -17,15 +19,17 @@ def initialize(saml_response, user, session)
       @context = nil
     end
 
-    def known_error?
-      known_errors = [
-        :clicked_deny?,
-        :auth_too_late?,
-        :auth_too_early?
-      ]
+    def error
+      KNOWN_ERRORS.each do |known_error|
+        return known_error if send("#{known_error}?")
+      end
```

Comment:
Do you mean [Hash#invert](https://ruby-doc.org/core-2.3.0/Hash.html#method-i-invert)?  Hash#transpose does not exist.

A few problems though, one is that the JSON error can exist in 2 places:
- `@saml_response.errors` array
- `@saml_response.status_message`

Second, the error message for too early & too late is on a substring.  To put so the hash would have to look like this:
```ruby
KNOWN_ERRORS = {
  auth_too_late:  'Current time is on'
  auth_too_early: 'Current time is ea'
}.freeze
```
 It starts to look hairy.